### ISSUE TITLE
bug in regexp m/(.*?)b/ig

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7080]
Migrated from [rt.perl.org#25822](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=25822) (status was 'resolved')

Searchable as RT25822$

**Comment by p5pRT** [comment]
### From chatolog@no4.dca1.superb.net

#### Created by daniel@chatologica.com



The following perl code uses too much cpu resources on perl 5\.8\.0 and
5\.8\.2\. I have not tested this on perl 5\.8\.3\.

The code below will use 2\-3 cpu seconds\.
On perl 5\.6\.1 it will took only 20ms cpu usage\.

$str = "a" x 20000;
while\($str =~ m/\(\.\*?\)b/ig\)\{\};
print "CPU times&#8203;:" \. times;

If you remove brackets \(\) or "i" modifier from the regexp code above the
cpu
usage will be normal\(low\)\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.0:

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.9, archname=sun4-solaris
    uname='sunos solaris 5.9 generic sun4u sparc sunw,ultra-5_10 '
    config_args='-Dcc=gcc -B/usr/ccs/bin/'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc -B/usr/ccs/bin/', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='3.1', gccosandvers='solaris2.9'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc -B/usr/ccs/bin/', ldflags =' -L/usr/local/lib '
    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/local/lib/perl5/5.8.0/sun4-solaris
    /usr/local/lib/perl5/5.8.0
    /usr/local/lib/perl5/site_perl/5.8.0/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home3/chatologica
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/usr/lib/nis/bin:/usr/local/bin:/usr/local/Hughes/bin:/bin:/usr/java/bin:/usr/local/pgsql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 03&#8203;:09 \+0000 2/2/04\, chatolog@&#8203;no4\.dca1\.superb\.net \(via RT\) wrote&#8203;:
>The code below will use 2\-3 cpu seconds\.
>On perl 5\.6\.1 it will took only 20ms cpu usage\.
>
>$str = "a" x 20000;
>while\($str =~ m/\(\.\*?\)b/ig\)\{\};
>print "CPU times&#8203;:" \. times;
>
>If you remove brackets \(\) or "i" modifier from the regexp code above the
>cpu
>usage will be normal\(low\)\.

I'm seeing the same with 5\.8\.3 and 5\.9\.0\.  And oddly\, I'm seeing a 
slightly \_lower\_ CPU usage in threaded Perl as opposed to unthreaded 
Perls \(very pronounced on Mac OS X\, and less so on Linux\)\.  I would 
have expected the threaded Perls to need more CPU\.


Liz

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Quoting Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl>&#8203;:

> At 03&#8203;:09 \+0000 2/2/04\, chatolog@&#8203;no4\.dca1\.superb\.net \(via RT\) wrote&#8203;:
> >The code below will use 2\-3 cpu seconds\.
> >On perl 5\.6\.1 it will took only 20ms cpu usage\.
> >
> >$str = "a" x 20000;
> >while\($str =~ m/\(\.\*?\)b/ig\)\{\};
> >print "CPU times&#8203;:" \. times;
> >
> >If you remove brackets \(\) or "i" modifier from the regexp code above the
> >cpu
> >usage will be normal\(low\)\.
> 
> I'm seeing the same with 5\.8\.3 and 5\.9\.0\.  And oddly\, I'm seeing a 
> slightly \_lower\_ CPU usage in threaded Perl as opposed to unthreaded 
> Perls \(very pronounced on Mac OS X\, and less so on Linux\)\.  I would 
> have expected the threaded Perls to need more CPU\.

If you add a "my" in front of "$str = \.\.\." you should have roughly
equivalent numbers\. Right?

**Comment by p5pRT** [comment]
### From @lizmat

At 11&#8203;:57 \+0100 2/3/04\, Rafael Garcia\-Suarez wrote&#8203;:
>Quoting Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl>&#8203;:
>  > At 03&#8203;:09 \+0000 2/2/04\, chatolog@&#8203;no4\.dca1\.superb\.net \(via RT\) wrote&#8203;:
>>  >The code below will use 2\-3 cpu seconds\.
>>  >On perl 5\.6\.1 it will took only 20ms cpu usage\.
>>  >
>  > >$str = "a" x 20000;
>>  >while\($str =~ m/\(\.\*?\)b/ig\)\{\};
>  > >print "CPU times&#8203;:" \. times;
>>  >
>>  >If you remove brackets \(\) or "i" modifier from the regexp code above the
>>  >cpu
>>  >usage will be normal\(low\)\.
>>
>>  I'm seeing the same with 5\.8\.3 and 5\.9\.0\.  And oddly\, I'm seeing a
>>  slightly \_lower\_ CPU usage in threaded Perl as opposed to unthreaded
>>  Perls \(very pronounced on Mac OS X\, and less so on Linux\)\.  I would
>  > have expected the threaded Perls to need more CPU\.
>If you add a "my" in front of "$str = \.\.\." you should have roughly
>equivalent numbers\. Right?

Nope\, same difference&#8203;:

$ perl5\.8\.3\-unthreaded 1
CPU times&#8203;:1\.73
$ perl5\.8\.3\-threaded 1
CPU times&#8203;:1\.59


Liz

**Comment by p5pRT** [comment]
### From @demerphq

On Sun Feb 01 19&#8203;:09&#8203;:46 2004\, chatolog@&#8203;no4\.dca1\.superb\.net wrote&#8203;:
> This is a bug report for perl from daniel@&#8203;chatologica\.com\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> The following perl code uses too much cpu resources on perl 5\.8\.0 and
> 5\.8\.2\. I have not tested this on perl 5\.8\.3\.
> 
> The code below will use 2\-3 cpu seconds\.
> On perl 5\.6\.1 it will took only 20ms cpu usage\.
> 
> $str = "a" x 20000;
> while\($str =~ m/\(\.\*?\)b/ig\)\{\};
> print "CPU times&#8203;:" \. times;
> 
> If you remove brackets \(\) or "i" modifier from the regexp code above
> the cpu usage will be normal\(low\)\.

Hi\,

Thanks for the report\. This was fixed in change \#29310 in Perl 5\.9\.5

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/29310

I have closed your ticket\.

Cheers\,
Yves
\*Please do not reply to this email unless you want to reopen the ticket\*

**Comment by p5pRT** [comment]
@demerphq - Status changed from 'open' to 'resolved'
